# Recommender system week 2

In [120]:
import pandas as pd
import numpy as np


In [121]:
data = pd.read_csv("HW1-data.csv")

In [122]:
data

User  Gender (1 =F, 0=M)  260: Star Wars: Episode IV - A New Hope (1977)  \
0    755                   0                                             1.0   
1   5277                   0                                             5.0   
2   1577                   1                                             NaN   
3   4388                   0                                             NaN   
4   1202                   1                                             4.0   
5   3823                   1                                             2.0   
6   5448                   0                                             NaN   
7   5347                   0                                             4.0   
8   4117                   1                                             5.0   
9   2765                   0                                             4.0   
10  5450                   1                                             2.0   
11   139                   0                                             3.0   
12  1940                   0                                             2.0   
13  3118                   1                                             3.0   
14  4656                   1                                             4.0   
15  4796                   1                                             NaN   
16  6037                   0                                             NaN   
17  3048                   1                                             4.0   
18  4790                   0                                             5.0   
19  4489                   0                                             1.0   

    1210: Star Wars: Episode VI - Return of the Jedi (1983)  \
0                                                 5.0         
1                                                 3.0         
2                                                 NaN         
3                                                 3.0         
4                                                 3.0         
5                                                 4.0         
6                                                 NaN         
7                                                 NaN         
8                                                 1.0         
9                                                 2.0         
10                                                1.0         
11                                                5.0         
12                                                3.0         
13                                                NaN         
14                                                4.0         
15                                                NaN         
16                                                NaN         
17                                                5.0         
18                                                1.0         
19                                                2.0         

    356: Forrest Gump (1994)  318: Shawshank Redemption, The (1994)  \
0                        2.0                                    NaN   
1                        NaN                                    2.0   
2                        NaN                                    5.0   
3                        NaN                                    NaN   
4                        4.0                                    1.0   
5                        4.0                                    4.0   
6                        3.0                                    1.0   
7                        NaN                                    NaN   
8                        NaN                                    4.0   
9                        NaN                                    5.0   
10                       5.0                                    NaN   
11                       2.0                                    NaN   
12                       NaN                                    5.0   
13   

# ((((( Mean )))))

Mean Rating: Calculate the mean rating for each movie, order with the highest rating listed first, and submit the top three (along with the mean scores for the top two).


In [123]:
 mean = data.agg(['mean'])

In [124]:
mean[:]

User  Gender (1 =F, 0=M)  \
mean  3658.1                0.45   

      260: Star Wars: Episode IV - A New Hope (1977)  \
mean                                        3.266667   

      1210: Star Wars: Episode VI - Return of the Jedi (1983)  \
mean                                                3.0         

      356: Forrest Gump (1994)  318: Shawshank Redemption, The (1994)  \
mean                       2.7                                    3.6   

      593: Silence of the Lambs, The (1991)  3578: Gladiator (2000)  \
mean                                 3.0625                2.916667   

      1: Toy Story (1995)  2028: Saving Private Ryan (1998)  ...  \
mean             2.823529                               3.0  ...   

      2396: Shakespeare in Love (1998)  2916: Total Recall (1990)  \
mean                          2.909091                   1.916667   

      780: Independence Day (ID4) (1996)  541: Blade Runner (1982)  \
mean                            2.769231                  3.222222   

      1265: Groundhog Day (1993)  2571: Matrix, The (1999)  \
mean                    3.166667                  2.833333   

      527: Schindler's List (1993)  2762: Sixth Sense, The (1999)  \
mean                           3.0                       2.833333   

      1198: Raiders of the Lost Ark (1981)  34: Babe (1995)  
mean                              2.909091              3.0  

[1 rows x 22 columns]

In [125]:
mean.drop(['Gender (1 =F, 0=M)','User'],axis = 1 , inplace = True)

In [126]:
mean

260: Star Wars: Episode IV - A New Hope (1977)  \
mean                                        3.266667   

      1210: Star Wars: Episode VI - Return of the Jedi (1983)  \
mean                                                3.0         

      356: Forrest Gump (1994)  318: Shawshank Redemption, The (1994)  \
mean                       2.7                                    3.6   

      593: Silence of the Lambs, The (1991)  3578: Gladiator (2000)  \
mean                                 3.0625                2.916667   

      1: Toy Story (1995)  2028: Saving Private Ryan (1998)  \
mean             2.823529                               3.0   

      296: Pulp Fiction (1994)  1259: Stand by Me (1986)  \
mean                       3.0                  2.636364   

      2396: Shakespeare in Love (1998)  2916: Total Recall (1990)  \
mean                          2.909091                   1.916667   

      780: Independence Day (ID4) (1996)  541: Blade Runner (1982)  \
mean                            2.769231                  3.222222   

      1265: Groundhog Day (1993)  2571: Matrix, The (1999)  \
mean                    3.166667                  2.833333   

      527: Schindler's List (1993)  2762: Sixth Sense, The (1999)  \
mean                           3.0                       2.833333   

      1198: Raiders of the Lost Ark (1981)  34: Babe (1995)  
mean                              2.909091              3.0

In [127]:
mean.sort_values(by = 'mean', axis=1, ascending=False, inplace=False, kind='quicksort', na_position='last')

318: Shawshank Redemption, The (1994)  \
mean                                    3.6   

      260: Star Wars: Episode IV - A New Hope (1977)  \
mean                                        3.266667   

      541: Blade Runner (1982)  1265: Groundhog Day (1993)  \
mean                  3.222222                    3.166667   

      593: Silence of the Lambs, The (1991)  296: Pulp Fiction (1994)  \
mean                                 3.0625                       3.0   

      527: Schindler's List (1993)  \
mean                           3.0   

      1210: Star Wars: Episode VI - Return of the Jedi (1983)  \
mean                                                3.0         

      34: Babe (1995)  2028: Saving Private Ryan (1998)  \
mean              3.0                               3.0   

      3578: Gladiator (2000)  1198: Raiders of the Lost Ark (1981)  \
mean                2.916667                              2.909091   

      2396: Shakespeare in Love (1998)  2571: Matrix, The (1999)  \
mean                          2.909091                  2.833333   

      2762: Sixth Sense, The (1999)  1: Toy Story (1995)  \
mean                       2.833333             2.823529   

      780: Independence Day (ID4) (1996)  356: Forrest Gump (1994)  \
mean                            2.769231                       2.7   

      1259: Stand by Me (1986)  2916: Total Recall (1990)  
mean                  2.636364                   1.916667

# check 1

In [128]:
mean['356: Forrest Gump (1994)']

mean    2.7
Name: 356: Forrest Gump (1994), dtype: float64

# 2 -(((((( Count )))))))

Rating Count (popularity): Count the number of ratings for each movie, order with the most number of ratings first, and submit the top three (along with the counts for the top two).


In [129]:
 count = data.agg(['count'])

In [130]:
count

User  Gender (1 =F, 0=M)  \
count    20                  20   

       260: Star Wars: Episode IV - A New Hope (1977)  \
count                                              15   

       1210: Star Wars: Episode VI - Return of the Jedi (1983)  \
count                                                 14         

       356: Forrest Gump (1994)  318: Shawshank Redemption, The (1994)  \
count                        10                                     10   

       593: Silence of the Lambs, The (1991)  3578: Gladiator (2000)  \
count                                     16                      12   

       1: Toy Story (1995)  2028: Saving Private Ryan (1998)  ...  \
count                   17                                11  ...   

       2396: Shakespeare in Love (1998)  2916: Total Recall (1990)  \
count                                11                         12   

       780: Independence Day (ID4) (1996)  541: Blade Runner (1982)  \
count                                  13                         9   

       1265: Groundhog Day (1993)  2571: Matrix, The (1999)  \
count                          12                        12   

       527: Schindler's List (1993)  2762: Sixth Sense, The (1999)  \
count                            12                             12   

       1198: Raiders of the Lost Ark (1981)  34: Babe (1995)  
count                                    11               10  

[1 rows x 22 columns]

In [131]:
count.sort_values(by = 'count', axis=1, ascending=False, inplace=False, kind='quicksort', na_position='last')

User  Gender (1 =F, 0=M)  1: Toy Story (1995)  \
count    20                  20                   17   

       593: Silence of the Lambs, The (1991)  \
count                                     16   

       260: Star Wars: Episode IV - A New Hope (1977)  \
count                                              15   

       1210: Star Wars: Episode VI - Return of the Jedi (1983)  \
count                                                 14         

       780: Independence Day (ID4) (1996)  2762: Sixth Sense, The (1999)  \
count                                  13                             12   

       3578: Gladiator (2000)  527: Schindler's List (1993)  ...  \
count                      12                            12  ...   

       2916: Total Recall (1990)  1198: Raiders of the Lost Ark (1981)  \
count                         12                                    11   

       1259: Stand by Me (1986)  2396: Shakespeare in Love (1998)  \
count                        11                                11   

       296: Pulp Fiction (1994)  2028: Saving Private Ryan (1998)  \
count                        11                                11   

       318: Shawshank Redemption, The (1994)  356: Forrest Gump (1994)  \
count                                     10                        10   

       34: Babe (1995)  541: Blade Runner (1982)  
count               10                         9  

[1 rows x 22 columns]

# check 2

In [132]:
count['356: Forrest Gump (1994)']

count    10
Name: 356: Forrest Gump (1994), dtype: int64

# 3 - (((((( > 4))))))

% of ratings 4+ (liking): Calculate the percentage of ratings for each movie that are 4 or higher. Order with the highest percentage first, and submit the to
p three (along with the percentage for the top two). Notice that the three different measures of "best" reflect different priorities and give different results; this should help you see why you need to be thoughtful about what metrics you use.

In [133]:
data_3 = data.drop(['User','Gender (1 =F, 0=M)'],axis = 1 )

In [134]:
data_3

260: Star Wars: Episode IV - A New Hope (1977)  \
0                                              1.0   
1                                              5.0   
2                                              NaN   
3                                              NaN   
4                                              4.0   
5                                              2.0   
6                                              NaN   
7                                              4.0   
8                                              5.0   
9                                              4.0   
10                                             2.0   
11                                             3.0   
12                                             2.0   
13                                             3.0   
14                                             4.0   
15                                             NaN   
16                                             NaN   
17                                             4.0   
18                                             5.0   
19                                             1.0   

    1210: Star Wars: Episode VI - Return of the Jedi (1983)  \
0                                                 5.0         
1                                                 3.0         
2                                                 NaN         
3                                                 3.0         
4                                                 3.0         
5                                                 4.0         
6                                                 NaN         
7                                                 NaN         
8                                                 1.0         
9                                                 2.0         
10                                                1.0         
11                                                5.0         
12                                                3.0         
13                                                NaN         
14                                                4.0         
15                                                NaN         
16                                                NaN         
17                                                5.0         
18                                                1.0         
19                                                2.0         

    356: Forrest Gump (1994)  318: Shawshank Redemption, The (1994)  \
0                        2.0                                    NaN   
1                        NaN                                    2.0   
2                        NaN                                    5.0   
3                        NaN                                    NaN   
4                        4.0                                    1.0   
5                        4.0                                    4.0   
6                        3.0                                    1.0   
7                        NaN                                    NaN   
8                        NaN                                    4.0   
9                        NaN                                    5.0   
10                       5.0                                    NaN   
11                       2.0                                    NaN   
12                       NaN                                    5.0   
13                       3.0                                    NaN   
14                       NaN                                    NaN   
15                       1.0                                    NaN   
16                       NaN                                    NaN   
17                       1.0                                    5.0   
18                       NaN                                    NaN   
19                       2.0                                    4.0   

    593: Silence of the Lambs, The (1991)  3578: Glad

In [135]:
def ofgreaterthan4(x):
    if (x >= 4):
        return 1
    else :
        return 0 
    
cols = data_3.columns    

In [136]:
data1 = data_3.copy()

In [137]:
for col in cols :
    data1[col] = data1[col].apply(lambda x: ofgreaterthan4(x))

In [138]:
data1

260: Star Wars: Episode IV - A New Hope (1977)  \
0                                                0   
1                                                1   
2                                                0   
3                                                0   
4                                                1   
5                                                0   
6                                                0   
7                                                1   
8                                                1   
9                                                1   
10                                               0   
11                                               0   
12                                               0   
13                                               0   
14                                               1   
15                                               0   
16                                               0   
17                                               1   
18                                               1   
19                                               0   

    1210: Star Wars: Episode VI - Return of the Jedi (1983)  \
0                                                   1         
1                                                   0         
2                                                   0         
3                                                   0         
4                                                   0         
5                                                   1         
6                                                   0         
7                                                   0         
8                                                   0         
9                                                   0         
10                                                  0         
11                                                  1         
12                                                  0         
13                                                  0         
14                                                  1         
15                                                  0         
16                                                  0         
17                                                  1         
18                                                  0         
19                                                  0         

    356: Forrest Gump (1994)  318: Shawshank Redemption, The (1994)  \
0                          0                                      0   
1                          0                                      0   
2                          0                                      1   
3                          0                                      0   
4                          1                                      0   
5                          1                                      1   
6                          0                                      0   
7                          0                                      0   
8                          0                                      1   
9                          0                                      1   
10                         1                                      0   
11                         0                                      0   
12                         0                                      1   
13                         0                                      0   
14                         0                                      0   
15                         0                                      0   
16                         0                                      0   
17                         0                                      1   
18                         0                                      0   
19                         0                                      1   

    593: Silence of the Lambs, The (1991)  3578: Glad

In [139]:
Sum = data1.sum()
Count = data_3.count()
Liking = Sum/Count

In [140]:
Liking.sort_values(ascending=False, inplace=False, kind='quicksort', na_position='last')

318: Shawshank Redemption, The (1994)                      0.700000
260: Star Wars: Episode IV - A New Hope (1977)             0.533333
3578: Gladiator (2000)                                     0.500000
541: Blade Runner (1982)                                   0.444444
593: Silence of the Lambs, The (1991)                      0.437500
2571: Matrix, The (1999)                                   0.416667
1265: Groundhog Day (1993)                                 0.416667
34: Babe (1995)                                            0.400000
296: Pulp Fiction (1994)                                   0.363636
2028: Saving Private Ryan (1998)                           0.363636
1259: Stand by Me (1986)                                   0.363636
1210: Star Wars: Episode VI - Return of the Jedi (1983)    0.357143
1: Toy Story (1995)                                        0.352941
527: Schindler's List (1993)                               0.333333
2762: Sixth Sense, The (1999)                   

# check 3 

In [141]:
Liking['356: Forrest Gump (1994)']

0.3

# 4 -  (((((((( Toy story ))))))))

Top movies for someone who has seen Toy Story: Calculate movies that most often occur with Movie #1: Toy Story, using the (x+y)/x method described in class. In other words, for each movie, calculate the percentage of Toy Story raters who also rated that movie. Order with the highest percentage first, and submit the top 3 (along with the correlations for the top two). Note, you will have ties -- to break the ties, use the lowest-numbered movie as the higher-ranked one. In other words, if Movies 541 and 318 are tied, then 318 gets the higher rank.

In [142]:
data_4 = data[data['1: Toy Story (1995)'].notna()]


In [143]:
data_4

User  Gender (1 =F, 0=M)  260: Star Wars: Episode IV - A New Hope (1977)  \
0    755                   0                                             1.0   
1   5277                   0                                             5.0   
2   1577                   1                                             NaN   
3   4388                   0                                             NaN   
5   3823                   1                                             2.0   
7   5347                   0                                             4.0   
8   4117                   1                                             5.0   
9   2765                   0                                             4.0   
10  5450                   1                                             2.0   
11   139                   0                                             3.0   
12  1940                   0                                             2.0   
13  3118                   1                                             3.0   
14  4656                   1                                             4.0   
16  6037                   0                                             NaN   
17  3048                   1                                             4.0   
18  4790                   0                                             5.0   
19  4489                   0                                             1.0   

    1210: Star Wars: Episode VI - Return of the Jedi (1983)  \
0                                                 5.0         
1                                                 3.0         
2                                                 NaN         
3                                                 3.0         
5                                                 4.0         
7                                                 NaN         
8                                                 1.0         
9                                                 2.0         
10                                                1.0         
11                                                5.0         
12                                                3.0         
13                                                NaN         
14                                                4.0         
16                                                NaN         
17                                                5.0         
18                                                1.0         
19                                                2.0         

    356: Forrest Gump (1994)  318: Shawshank Redemption, The (1994)  \
0                        2.0                                    NaN   
1                        NaN                                    2.0   
2                        NaN                                    5.0   
3                        NaN                                    NaN   
5                        4.0                                    4.0   
7                        NaN                                    NaN   
8                        NaN                                    4.0   
9                        NaN                                    5.0   
10                       5.0                                    NaN   
11                       2.0                                    NaN   
12                       NaN                                    5.0   
13                       3.0                                    NaN   
14                       NaN                                    NaN   
16                       NaN                                    NaN   
17                       1.0                                    5.0   
18                       NaN                                    NaN   
19                       2.0                                    4.0   

    593: Silence of the Lambs, The (1991)  3578: Gladiator (2000)  \
0                                     4.0                     4.0   
1          

In [144]:
cols = data_4.columns 
ls =[]
for col in cols :
    perc = data_4[col].count() /  data_4['1: Toy Story (1995)'].count()
    print("Perc of ",col, "is", perc)
    ls.append(perc)
    

    
ls.sort(reverse=True)
print("---------------------")
print("List is ")
print(ls)

Perc of  User is 1.0
Perc of  Gender (1 =F, 0=M) is 1.0
Perc of  260: Star Wars: Episode IV - A New Hope (1977) is 0.8235294117647058
Perc of  1210: Star Wars: Episode VI - Return of the Jedi (1983) is 0.7647058823529411
Perc of  356: Forrest Gump (1994) is 0.4117647058823529
Perc of  318: Shawshank Redemption, The (1994) is 0.47058823529411764
Perc of  593: Silence of the Lambs, The (1991) is 0.7647058823529411
Perc of  3578: Gladiator (2000) is 0.5294117647058824
Perc of  1: Toy Story (1995) is 1.0
Perc of  2028: Saving Private Ryan (1998) is 0.47058823529411764
Perc of  296: Pulp Fiction (1994) is 0.5882352941176471
Perc of  1259: Stand by Me (1986) is 0.5294117647058824
Perc of  2396: Shakespeare in Love (1998) is 0.47058823529411764
Perc of  2916: Total Recall (1990) is 0.6470588235294118
Perc of  780: Independence Day (ID4) (1996) is 0.7647058823529411
Perc of  541: Blade Runner (1982) is 0.4117647058823529
Perc of  1265: Groundhog Day (1993) is 0.6470588235294118
Perc of  2571: 

# check 4 

Perc of  356: Forrest Gump (1994) is 


0.4117647058823529

# 5 - ((((((corr))))))

In [145]:
print(data_4.corr()[u'1: Toy Story (1995)'].sort_values(ascending=False).head(6))


1: Toy Story (1995)                      1.000000
318: Shawshank Redemption, The (1994)    0.888523
34: Babe (1995)                          0.811107
296: Pulp Fiction (1994)                 0.709842
2028: Saving Private Ryan (1998)         0.596849
Gender (1 =F, 0=M)                       0.570232
Name: 1: Toy Story (1995), dtype: float64


# check 5 

356: Forrest Gump (1994)             

0.522913


# 6 - ((((((()))))))

Mean rating difference by gender: First, recompute the mean rating for each movie separately for males and for females. And calculate the overall mean rating (across all ratings) for males and females. Submit the two movies that have the greatest differences (one where men are most above women, and one where women are most above men) along with the differences in average. Also submit the difference in overall rating averages (female average - male average).

In [146]:
#check the data

#data

In [147]:
# mean all 

data_mean = data.agg(['mean']).drop(['User','Gender (1 =F, 0=M)'],axis = 1)
data_mean

260: Star Wars: Episode IV - A New Hope (1977)  \
mean                                        3.266667   

      1210: Star Wars: Episode VI - Return of the Jedi (1983)  \
mean                                                3.0         

      356: Forrest Gump (1994)  318: Shawshank Redemption, The (1994)  \
mean                       2.7                                    3.6   

      593: Silence of the Lambs, The (1991)  3578: Gladiator (2000)  \
mean                                 3.0625                2.916667   

      1: Toy Story (1995)  2028: Saving Private Ryan (1998)  \
mean             2.823529                               3.0   

      296: Pulp Fiction (1994)  1259: Stand by Me (1986)  \
mean                       3.0                  2.636364   

      2396: Shakespeare in Love (1998)  2916: Total Recall (1990)  \
mean                          2.909091                   1.916667   

      780: Independence Day (ID4) (1996)  541: Blade Runner (1982)  \
mean                            2.769231                  3.222222   

      1265: Groundhog Day (1993)  2571: Matrix, The (1999)  \
mean                    3.166667                  2.833333   

      527: Schindler's List (1993)  2762: Sixth Sense, The (1999)  \
mean                           3.0                       2.833333   

      1198: Raiders of the Lost Ark (1981)  34: Babe (1995)  
mean                              2.909091              3.0

In [148]:
data_males = data[data['Gender (1 =F, 0=M)']== 0]
data_females = data[data['Gender (1 =F, 0=M)']== 1]

In [149]:
data_females

User  Gender (1 =F, 0=M)  260: Star Wars: Episode IV - A New Hope (1977)  \
2   1577                   1                                             NaN   
4   1202                   1                                             4.0   
5   3823                   1                                             2.0   
8   4117                   1                                             5.0   
10  5450                   1                                             2.0   
13  3118                   1                                             3.0   
14  4656                   1                                             4.0   
15  4796                   1                                             NaN   
17  3048                   1                                             4.0   

    1210: Star Wars: Episode VI - Return of the Jedi (1983)  \
2                                                 NaN         
4                                                 3.0         
5                                                 4.0         
8                                                 1.0         
10                                                1.0         
13                                                NaN         
14                                                4.0         
15                                                NaN         
17                                                5.0         

    356: Forrest Gump (1994)  318: Shawshank Redemption, The (1994)  \
2                        NaN                                    5.0   
4                        4.0                                    1.0   
5                        4.0                                    4.0   
8                        NaN                                    4.0   
10                       5.0                                    NaN   
13                       3.0                                    NaN   
14                       NaN                                    NaN   
15                       1.0                                    NaN   
17                       1.0                                    5.0   

    593: Silence of the Lambs, The (1991)  3578: Gladiator (2000)  \
2                                     2.0                     NaN   
4                                     4.0                     1.0   
5                                     NaN                     NaN   
8                                     2.0                     4.0   
10                                    NaN                     5.0   
13                                    2.0                     NaN   
14                                    5.0                     5.0   
15                                    3.0                     2.0   
17                                    1.0                     1.0   

    1: Toy Story (1995)  2028: Saving Private Ryan (1998)  ...  \
2                   4.0                               NaN  ...   
4                   NaN                               4.0  ...   
5                   3.0                               1.0  ...   
8                   4.0                               4.0  ...   
10                  5.0                               NaN  ...   
13                  3.0                               NaN  ...   
14                  2.0                               NaN  ...   
15                  NaN                               2.0  ...   
17                  4.0                               NaN  ...   

    2396: Shakespeare in Love (1998)  2916: Total Recall (1990)  \
2                                NaN                        1.0   
4                                5.0                        1.0   
5                                5.0                        2.0   
8                                2.0                        3.0   
10                               NaN                        3.0   
13                               NaN                        1.0   
14                               NaN    

In [150]:
# diff_= []
# for col in cols:
#     diff = data_f[col].item() - data_m[col].item()
#     print("diff for ",col,"is : ", diff)
#     diff_.append(diff)
# diff_.sort()
# print(diff_)

where males are the most : diff for  1198: Raiders of the Lost Ark (1981) is :  -1.6666666666666665



where females are the most :  diff for  2396: Shakespeare in Love (1998) is :  2.107142857142857


In [151]:
2.107142857142857 - 1.6666666666666665

0.4404761904761907

# Check 6 

In [152]:
data_f = data_females.agg(['mean'])
data_f['356: Forrest Gump (1994)']

mean    3.0
Name: 356: Forrest Gump (1994), dtype: float64

In [153]:
data_m = data_males.agg(['mean'])
data_m['356: Forrest Gump (1994)']

mean    2.25
Name: 356: Forrest Gump (1994), dtype: float64

 diff for  356: Forrest Gump (1994) is :  0.75


# 7 - (((((())))))

Next, compute the % of ratings 4+ separately for males and females. You'll be asked to submit two movies as above (largest difference in each direction). And again you'll indicate whether men or women are more likely to rate movies 4 stars or above.


In [154]:
# for females
data_females_ = data_females.copy()
data_males_ = data_males.copy()
for col in cols :
    data_females_[col] = data_females_[col].apply(lambda x: ofgreaterthan4(x))

# for males

for col in cols :
    data_males_[col] = data_males_[col].apply(lambda x: ofgreaterthan4(x))

In [155]:
data_males_

User  Gender (1 =F, 0=M)  260: Star Wars: Episode IV - A New Hope (1977)  \
0      1                   0                                               0   
1      1                   0                                               1   
3      1                   0                                               0   
6      1                   0                                               0   
7      1                   0                                               1   
9      1                   0                                               1   
11     1                   0                                               0   
12     1                   0                                               0   
16     1                   0                                               0   
18     1                   0                                               1   
19     1                   0                                               0   

    1210: Star Wars: Episode VI - Return of the Jedi (1983)  \
0                                                   1         
1                                                   0         
3                                                   0         
6                                                   0         
7                                                   0         
9                                                   0         
11                                                  1         
12                                                  0         
16                                                  0         
18                                                  0         
19                                                  0         

    356: Forrest Gump (1994)  318: Shawshank Redemption, The (1994)  \
0                          0                                      0   
1                          0                                      0   
3                          0                                      0   
6                          0                                      0   
7                          0                                      0   
9                          0                                      1   
11                         0                                      0   
12                         0                                      1   
16                         0                                      0   
18                         0                                      0   
19                         0                                      1   

    593: Silence of the Lambs, The (1991)  3578: Gladiator (2000)  \
0                                       1                       1   
1                                       1                       0   
3                                       0                       0   
6                                       0                       1   
7                                       0                       0   
9                                       0                       0   
11                                      0                       0   
12                                      1                       0   
16                                      1                       0   
18                                      0                       1   
19                                      1                       0   

    1: Toy Story (1995)  2028: Saving Private Ryan (1998)  ...  \
0                     0                                 0  ...   
1                     0                                 0  ...   
3                     0                                 0  ...   
6                     0                                 1  ...   
7                     0                                 0  ...   
9                     1                                 0  ...   
11                    0                                 0  ...   
12                    1                                 1  ...   
16        

In [156]:
# # And some oerall statistics:

# Average male rating: 2.906
# Percentage male ratings >=4: 0.3386

In [157]:
#for females 
Sum = data_females_.sum()
Count = data_females.count()
Liking_females = Sum/Count
print(Liking_females)

User                                                       1.000000
Gender (1 =F, 0=M)                                         0.000000
260: Star Wars: Episode IV - A New Hope (1977)             0.571429
1210: Star Wars: Episode VI - Return of the Jedi (1983)    0.500000
356: Forrest Gump (1994)                                   0.500000
318: Shawshank Redemption, The (1994)                      0.800000
593: Silence of the Lambs, The (1991)                      0.285714
3578: Gladiator (2000)                                     0.500000
1: Toy Story (1995)                                        0.571429
2028: Saving Private Ryan (1998)                           0.500000
296: Pulp Fiction (1994)                                   0.666667
1259: Stand by Me (1986)                                   0.428571
2396: Shakespeare in Love (1998)                           0.750000
2916: Total Recall (1990)                                  0.000000
780: Independence Day (ID4) (1996)              

In [160]:
Sum = data_males_.sum()
Count = data_males.count()
Liking_males = Sum/Count
print(Liking_males)
Liking_males[2]


User                                                       1.000000
Gender (1 =F, 0=M)                                         0.000000
260: Star Wars: Episode IV - A New Hope (1977)             0.500000
1210: Star Wars: Episode VI - Return of the Jedi (1983)    0.250000
356: Forrest Gump (1994)                                   0.000000
318: Shawshank Redemption, The (1994)                      0.600000
593: Silence of the Lambs, The (1991)                      0.555556
3578: Gladiator (2000)                                     0.500000
1: Toy Story (1995)                                        0.200000
2028: Saving Private Ryan (1998)                           0.285714
296: Pulp Fiction (1994)                                   0.250000
1259: Stand by Me (1986)                                   0.250000
2396: Shakespeare in Love (1998)                           0.000000
2916: Total Recall (1990)                                  0.200000
780: Independence Day (ID4) (1996)              

0.5

In [165]:
diff__= []
for col in cols:
    diff = Liking_females[col].item() - Liking_males[col].item()
    print("diff for ",col,"is : ", diff)
    diff__.append(diff)
diff__.sort()
print(diff__)

diff for  User is :  0.0
diff for  Gender (1 =F, 0=M) is :  0.0
diff for  260: Star Wars: Episode IV - A New Hope (1977) is :  0.0714285714285714
diff for  1210: Star Wars: Episode VI - Return of the Jedi (1983) is :  0.25
diff for  356: Forrest Gump (1994) is :  0.5
diff for  318: Shawshank Redemption, The (1994) is :  0.20000000000000007
diff for  593: Silence of the Lambs, The (1991) is :  -0.2698412698412699
diff for  3578: Gladiator (2000) is :  0.0
diff for  1: Toy Story (1995) is :  0.3714285714285714
diff for  2028: Saving Private Ryan (1998) is :  0.2142857142857143
diff for  296: Pulp Fiction (1994) is :  0.41666666666666663
diff for  1259: Stand by Me (1986) is :  0.17857142857142855
diff for  2396: Shakespeare in Love (1998) is :  0.75
diff for  2916: Total Recall (1990) is :  -0.2
diff for  780: Independence Day (ID4) (1996) is :  0.047619047619047616
diff for  541: Blade Runner (1982) is :  0.55
diff for  1265: Groundhog Day (1993) is :  -0.16666666666666669
diff for  257

# answer 

diff for  1198: Raiders of the Lost Ark (1981) is :  -0.5

Females

diff for  2396: Shakespeare in Love (1998) is :  0.75


        
        

=======================================


And some overall statistics:

Average male rating: 2.906
    
    
Percentage male ratings >=4: 0.3386

In [169]:
data_females_.drop(['User','Gender (1 =F, 0=M)'],axis = 1,inplace = True)
data_females_

260: Star Wars: Episode IV - A New Hope (1977)  \
2                                                0   
4                                                1   
5                                                0   
8                                                1   
10                                               0   
13                                               0   
14                                               1   
15                                               0   
17                                               1   

    1210: Star Wars: Episode VI - Return of the Jedi (1983)  \
2                                                   0         
4                                                   0         
5                                                   1         
8                                                   0         
10                                                  0         
13                                                  0         
14                                                  1         
15                                                  0         
17                                                  1         

    356: Forrest Gump (1994)  318: Shawshank Redemption, The (1994)  \
2                          0                                      1   
4                          1                                      0   
5                          1                                      1   
8                          0                                      1   
10                         1                                      0   
13                         0                                      0   
14                         0                                      0   
15                         0                                      0   
17                         0                                      1   

    593: Silence of the Lambs, The (1991)  3578: Gladiator (2000)  \
2                                       0                       0   
4                                       1                       0   
5                                       0                       0   
8                                       0                       1   
10                                      0                       1   
13                                      0                       0   
14                                      1                       1   
15                                      0                       0   
17                                      0                       0   

    1: Toy Story (1995)  2028: Saving Private Ryan (1998)  \
2                     1                                 0   
4                     0                                 1   
5                     0                                 0   
8                     1                                 1   
10                    1                                 0   
13                    0                                 0   
14                    0                                 0   
15                    0                                 0   
17                    1                                 0   

    296: Pulp Fiction (1994)  1259: Stand by Me (1986)  \
2                          0                         0   
4                          0                         0   
5                          1                         1   
8                          0                         0   
10                         0                         0   
13                         0                         1   
14                         0                         1   
15                         0                         0   
17                         1                         0   

    2396: Shakespeare in Love (1998)  2916: Total Recall (1990)  \
2                                  0                          0   
4                                  1                          0   
5                                  1 

In [187]:
data_females.drop(['User','Gender (1 =F, 0=M)'],axis = 1 , inplace = True)
data_females

260: Star Wars: Episode IV - A New Hope (1977)  \
2                                              NaN   
4                                              4.0   
5                                              2.0   
8                                              5.0   
10                                             2.0   
13                                             3.0   
14                                             4.0   
15                                             NaN   
17                                             4.0   

    1210: Star Wars: Episode VI - Return of the Jedi (1983)  \
2                                                 NaN         
4                                                 3.0         
5                                                 4.0         
8                                                 1.0         
10                                                1.0         
13                                                NaN         
14                                                4.0         
15                                                NaN         
17                                                5.0         

    356: Forrest Gump (1994)  318: Shawshank Redemption, The (1994)  \
2                        NaN                                    5.0   
4                        4.0                                    1.0   
5                        4.0                                    4.0   
8                        NaN                                    4.0   
10                       5.0                                    NaN   
13                       3.0                                    NaN   
14                       NaN                                    NaN   
15                       1.0                                    NaN   
17                       1.0                                    5.0   

    593: Silence of the Lambs, The (1991)  3578: Gladiator (2000)  \
2                                     2.0                     NaN   
4                                     4.0                     1.0   
5                                     NaN                     NaN   
8                                     2.0                     4.0   
10                                    NaN                     5.0   
13                                    2.0                     NaN   
14                                    5.0                     5.0   
15                                    3.0                     2.0   
17                                    1.0                     1.0   

    1: Toy Story (1995)  2028: Saving Private Ryan (1998)  \
2                   4.0                               NaN   
4                   NaN                               4.0   
5                   3.0                               1.0   
8                   4.0                               4.0   
10                  5.0                               NaN   
13                  3.0                               NaN   
14                  2.0                               NaN   
15                  NaN                               2.0   
17                  4.0                               NaN   

    296: Pulp Fiction (1994)  1259: Stand by Me (1986)  \
2                        NaN                       1.0   
4                        NaN                       1.0   
5                        4.0                       4.0   
8                        NaN                       1.0   
10                       NaN                       NaN   
13                       NaN                       4.0   
14                       3.0                       5.0   
15                       NaN                       1.0   
17                       5.0                       NaN   

    2396: Shakespeare in Love (1998)  2916: Total Recall (1990)  \
2                                NaN                        1.0   
4                                5.0                        1.0   
5                                5.0 

In [180]:
data_males.drop(['User','Gender (1 =F, 0=M)'],axis = 1,inplace = True)
data_males

260: Star Wars: Episode IV - A New Hope (1977)  \
0                                              1.0   
1                                              5.0   
3                                              NaN   
6                                              NaN   
7                                              4.0   
9                                              4.0   
11                                             3.0   
12                                             2.0   
16                                             NaN   
18                                             5.0   
19                                             1.0   

    1210: Star Wars: Episode VI - Return of the Jedi (1983)  \
0                                                 5.0         
1                                                 3.0         
3                                                 3.0         
6                                                 NaN         
7                                                 NaN         
9                                                 2.0         
11                                                5.0         
12                                                3.0         
16                                                NaN         
18                                                1.0         
19                                                2.0         

    356: Forrest Gump (1994)  318: Shawshank Redemption, The (1994)  \
0                        2.0                                    NaN   
1                        NaN                                    2.0   
3                        NaN                                    NaN   
6                        3.0                                    1.0   
7                        NaN                                    NaN   
9                        NaN                                    5.0   
11                       2.0                                    NaN   
12                       NaN                                    5.0   
16                       NaN                                    NaN   
18                       NaN                                    NaN   
19                       2.0                                    4.0   

    593: Silence of the Lambs, The (1991)  3578: Gladiator (2000)  \
0                                     4.0                     4.0   
1                                     4.0                     2.0   
3                                     NaN                     1.0   
6                                     1.0                     4.0   
7                                     3.0                     2.0   
9                                     3.0                     NaN   
11                                    2.0                     NaN   
12                                    4.0                     NaN   
16                                    4.0                     NaN   
18                                    NaN                     4.0   
19                                    5.0                     NaN   

    1: Toy Story (1995)  2028: Saving Private Ryan (1998)  \
0                   2.0                               2.0   
1                   1.0                               NaN   
3                   2.0                               3.0   
6                   NaN                               5.0   
7                   2.0                               NaN   
9                   4.0                               3.0   
11                  2.0                               NaN   
12                  4.0                               5.0   
16                  2.0                               NaN   
18                  2.0                               1.0   
19                  2.0                               3.0   

    296: Pulp Fiction (1994)  1259: Stand by Me (1986)  \
0                        NaN                       3.0   
1                        NaN                       4.0   
3        

In [174]:
data_females_

260: Star Wars: Episode IV - A New Hope (1977)  \
2                                                0   
4                                                1   
5                                                0   
8                                                1   
10                                               0   
13                                               0   
14                                               1   
15                                               0   
17                                               1   

    1210: Star Wars: Episode VI - Return of the Jedi (1983)  \
2                                                   0         
4                                                   0         
5                                                   1         
8                                                   0         
10                                                  0         
13                                                  0         
14                                                  1         
15                                                  0         
17                                                  1         

    356: Forrest Gump (1994)  318: Shawshank Redemption, The (1994)  \
2                          0                                      1   
4                          1                                      0   
5                          1                                      1   
8                          0                                      1   
10                         1                                      0   
13                         0                                      0   
14                         0                                      0   
15                         0                                      0   
17                         0                                      1   

    593: Silence of the Lambs, The (1991)  3578: Gladiator (2000)  \
2                                       0                       0   
4                                       1                       0   
5                                       0                       0   
8                                       0                       1   
10                                      0                       1   
13                                      0                       0   
14                                      1                       1   
15                                      0                       0   
17                                      0                       0   

    1: Toy Story (1995)  2028: Saving Private Ryan (1998)  \
2                     1                                 0   
4                     0                                 1   
5                     0                                 0   
8                     1                                 1   
10                    1                                 0   
13                    0                                 0   
14                    0                                 0   
15                    0                                 0   
17                    1                                 0   

    296: Pulp Fiction (1994)  1259: Stand by Me (1986)  \
2                          0                         0   
4                          0                         0   
5                          1                         1   
8                          0                         0   
10                         0                         0   
13                         0                         1   
14                         0                         1   
15                         0                         0   
17                         1                         0   

    2396: Shakespeare in Love (1998)  2916: Total Recall (1990)  \
2                                  0                          0   
4                                  1                          0   
5                                  1 

In [175]:
data_males_

User  Gender (1 =F, 0=M)  260: Star Wars: Episode IV - A New Hope (1977)  \
0      1                   0                                               0   
1      1                   0                                               1   
3      1                   0                                               0   
6      1                   0                                               0   
7      1                   0                                               1   
9      1                   0                                               1   
11     1                   0                                               0   
12     1                   0                                               0   
16     1                   0                                               0   
18     1                   0                                               1   
19     1                   0                                               0   

    1210: Star Wars: Episode VI - Return of the Jedi (1983)  \
0                                                   1         
1                                                   0         
3                                                   0         
6                                                   0         
7                                                   0         
9                                                   0         
11                                                  1         
12                                                  0         
16                                                  0         
18                                                  0         
19                                                  0         

    356: Forrest Gump (1994)  318: Shawshank Redemption, The (1994)  \
0                          0                                      0   
1                          0                                      0   
3                          0                                      0   
6                          0                                      0   
7                          0                                      0   
9                          0                                      1   
11                         0                                      0   
12                         0                                      1   
16                         0                                      0   
18                         0                                      0   
19                         0                                      1   

    593: Silence of the Lambs, The (1991)  3578: Gladiator (2000)  \
0                                       1                       1   
1                                       1                       0   
3                                       0                       0   
6                                       0                       1   
7                                       0                       0   
9                                       0                       0   
11                                      0                       0   
12                                      1                       0   
16                                      1                       0   
18                                      0                       1   
19                                      1                       0   

    1: Toy Story (1995)  2028: Saving Private Ryan (1998)  ...  \
0                     0                                 0  ...   
1                     0                                 0  ...   
3                     0                                 0  ...   
6                     0                                 1  ...   
7                     0                                 0  ...   
9                     1                                 0  ...   
11                    0                                 0  ...   
12                    1                                 1  ...   
16        

In [181]:
cols = data_males.columns

In [191]:
import math

count_rating4 = 0
count_rating_ = 0


for col in cols : 
    for element in data_males[col]:
        if element >= 4 :
            count_rating4 = count_rating4 +1
        elif element < 4 :
            count_rating_ = count_rating_ +1
        elif math.isnan(element):
            continue
#males rating greter than 4 
result_males = count_rating4 / (count_rating4+count_rating_)
            


# check 7 

In [195]:
result_males

0.33858267716535434

In [192]:
import math

count_rating4 = 0
count_rating_ = 0


for col in cols : 
    for element in data_females[col]:
        if element >= 4 :
            count_rating4 = count_rating4 +1
        elif element < 4 :
            count_rating_ = count_rating_ +1
        elif math.isnan(element):
            continue
#males rating greter than 4 
result_females = count_rating4 / (count_rating4+count_rating_)
            

In [193]:
result = result_females - result_males 

In [194]:
result

0.082469954413593